arch/x86/core/crt0.S

```diff
@@ -42,6 +42,84 @@
 	GTEXT(_sys_soc_resume_from_deep_sleep)
 #endif
 
+#ifdef CONFIG_JAILHOUSE
+
+#define JAILHOUSE_INMATE_CS32   0x8
+#define JAILHOUSE_INMATE_DS32   0x18
+#define MSR_MTRR_DEF_TYPE       0x000002ff
+#define MTRR_ENABLE             0x00000800
+#define X86_CR0_PE              0x00000001
+#define X86_CR0_WP              0x00010000
+#define X86_CR4_PSE             0x00000010
+
+	.code16
+	.section ".boot", "ax"
+
+	.globl __jh_entry
+__jh_entry:
+	ljmp $0x0,$start16
+
+	.section ".startup", "ax"
+start16:
+	lgdtl %cs:gdt_ptr
+
+	mov %cr0,%eax
+	or $X86_CR0_PE,%al
+	mov %eax,%cr0
+
+	ljmpl $JAILHOUSE_INMATE_CS32,$start32
+
+	.code32
+start32:
+	/*
+	 * Minimal bootstrap into 32-bit mode, just to jump to
+	 * __start
+	 */
+
+	/* 4Mb pages */
+	mov %cr4,%eax
+	or $X86_CR4_PSE,%eax
+	mov %eax,%cr4
+
+	mov $loader_pdpt,%eax
+	mov %eax,%cr3
+
+	/* Enable write protect and protected mode */
+	mov $(X86_CR0_WP | X86_CR0_PE),%eax
+	mov %eax,%cr0
+
+	movl $MSR_MTRR_DEF_TYPE,%ecx
```

Comment:
Is this something specific for Jailhouse?
Consider just moving this to __start

Reply:
I guess so, but correct me otherwise. Do Intel MCUs have this MSR as well? I guess this was put on the boot code to enable caching on simpler guests (the demos on Jailhouse), but I'm still confused because the manual says "When the E flag is set, the type specified in the default memory type field is used for areas of memory not already mapped by either a fixed or variable MTRR", and that type is *uncached*. @jan-kiszka , any input?

Reply:
This is IA32_MTRR_DEF_TYPE on Intel, and it is disabled on reset so that memory becomes uncached. Jailhouse hands the CPU over in reset state, therefore we have to enable it explicitly. On a normal system, the firmware tends to enable MTRRs already.

Reply:
ok, i'm fine with leaving this where it is